bip-0000.mediawiki

```diff
@@ -0,0 +1,274 @@
+
+* Let ''outpoints_hash = HASH(txid<sub>1</sub> || vout<sub>1</sub> || … txid<sub>n</sub> || vout<sub>n</sub>)''
+* Let ''a = a<sub>0</sub> + a<sub>1</sub> … + a<sub>n</sub>''
+* Let ''A = A<sub>0</sub> + A<sub>1</sub> … + A<sub>n</sub>''
+* Let ''P<sub>0</sub> = HASH(outpoints_hash·a·B || 0)·G + B''
+
+''' Spend and Scan Key '''
+
+Since Bob needs his private key ''b'' to check for incoming payments, this requires ''b'' to be exposed to an online device. To minimize the risks involved, Bob can instead publish an address of the form ''(B<sub>scan</sub>, B<sub>spend</sub>)''. This allows Bob to keep ''b<sub>spend</sub>'' in offline cold storage and perform the scanning with the public key ''B<sub>spend</sub>'' and private key ''b<sub>scan</sub>''. Alice performs the tweak using both of Bob’s public keys in the following manner:
+
+* Let ''P<sub>0</sub> = HASH(outpoints_hash·a·B<sub>scan</sub> || 0)·G + B<sub>spend</sub>''
+
+Bob detects this payment by calculating ''P<sub>0</sub> = HASH(outpoints_hash·b<sub>scan</sub>·A)·G + B<sub>spend</sub>'' with his online device and can spend from his cold storage signing device using ''(HASH(outpoints_hash·b<sub>scan</sub>·A) + b<sub>spend</sub>) mod p'' as the private key.
+
+''' Labels '''
+
+For a single silent payment address of the form ''(B<sub>scan</sub>, B<sub>spend</sub>)'', Bob may wish to differentiate incoming payments by using labels. Naively, Bob could publish multiple silent payment addresses, but this would require him to scan for each one, which becomes prohibitively expensive<ref name="labels">'''ECC multiplication vs ECC addition''' TODO</ref>. Instead, Bob can tweak his spend public key ''B<sub>spend</sub>'' with an integer ''m'' in the following way:
+
+* Let ''B<sub>m</sub> = B<sub>spend</sub> + m·G''
+* Publish ''(B<sub>scan</sub>, B<sub>0</sub>)'', ''(B<sub>scan</sub>, B<sub>1</sub>) …''
+
+Alice performs the tweak same as before using one of the published ''(B<sub>scan</sub>, B<sub>m</sub>)'' pairs. Bob detects the labeled payment in the following manner:
+
+* Let ''P<sub>0</sub> = HASH(outpoints_hash·b<sub>scan</sub>·A || 0)·G + B<sub>spend</sub>''
+* Compute ''P<sub>0m</sub> = P<sub>0</sub> + m·G'' for each ''m''
+* For each ''P<sub>0m</sub>'' in {''P<sub>00</sub> … P<sub>0m</sub>''}, check if any are present in the transaction outputs
+
+It is important to note that an outside observer can easily deduce that each published ''(B<sub>scan</sub>, B<sub>m</sub>)'' pair is owned by the same entity as each published address will have ''B<sub>scan</sub>'' in common. As such, labels are not meant as a way for Bob to manage separate identities, but rather a way for Bob to determine the source of an incoming payment.
+
+== Specification ==
+
+We use the following functions and conventions:
+
+* ser<sub>32</sub>(i): serializes a 32-bit unsigned integer i as a 4-byte sequence, most significant byte first
```

Comment:
For ease of implementation, I'd advocate for using little-endian (or "least significant byte first") byte order here instead, so we can use our internal serialization framework and avoid creating temporary buffers. For example, the outpoint hashing could then just serialize a COutPoint instance as-is:
```
COutPoint outpoint = ...;
HashWriter h;
h << outpoint;
uint256 outpoint_hash = h.GetSHA256();
```
With big endian, this would be more tedious and look something like
```
COutPoint outpoint = ...;
char outpoint_index_buf[4];
WriteBE32(outpoint_index_buf, outpoint.n);
HashWriter h;
h << outpoint.hash;
h.write(AsBytes(Span{outpoint_index_buf, 4}));
uint256 outpoint_hash = h.GetSHA256();

```
(The existing draft implementation is currently not respecting the BIP w.r.t. to outpoint serialization anyway, as it serializes the outpoint index with a 32-bytes instead of a 4-bytes sequence; that probably resulted out of bits/bytes confusion)

Reply:
great catch, thanks!